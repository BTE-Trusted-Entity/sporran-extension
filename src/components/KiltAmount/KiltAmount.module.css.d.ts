export const __esModule: true;
export const amount: string;
export const fractional: string;
