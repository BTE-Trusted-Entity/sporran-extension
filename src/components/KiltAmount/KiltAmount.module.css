.amount {
  composes: balanceText from '../Typography/Typography.module.css';
  margin-left: 0.25em;
  font-size: 1em;
  line-height: 1em;
}

.fractional {
  font-size: 0.7em;
  vertical-align: text-top;
}
