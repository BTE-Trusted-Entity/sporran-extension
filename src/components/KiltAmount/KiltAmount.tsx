import BN from 'bn.js';
import { browser } from 'webextension-polyfill-ts';

import * as styles from './KiltAmount.module.css';

import { KiltCurrency } from '../KiltCurrency/KiltCurrency';

const KILT_POWER = 15;

const PRECISION = 4;
const FORMAT = {
  minimumFractionDigits: PRECISION,
  maximumFractionDigits: PRECISION,
};

type Type = 'funds' | 'costs';

export function asKiltCoins(amount: BN, type: Type): string {
  const amountString = amount.toString().padStart(16, '0');
  const whole = amountString.slice(0, -KILT_POWER);
  const fraction = amountString.slice(-KILT_POWER);

  const visible = fraction.slice(0, PRECISION);
  const discarded = fraction.slice(PRECISION);
  const noRoundingNeeded = discarded.match(/^0*$/);
  const roundDown = noRoundingNeeded || type === 'funds';
  const replacement = roundDown ? '0' : '9';

  const numberWithFractions = parseFloat(`${whole}.${visible}${replacement}`);

  return numberWithFractions.toLocaleString(
    browser.i18n.getUILanguage(),
    FORMAT,
  );
}

interface Props {
  amount: BN;
  type: Type;
  smallDecimals?: boolean;
}

export function KiltAmount({
  amount,
  type,
  smallDecimals = false,
}: Props): JSX.Element {
  const value = asKiltCoins(amount, type);

  const offset = -1 - PRECISION;
  const integer = value.slice(0, offset);
  const fractional = value.slice(offset);

  const formatted = smallDecimals ? (
    <>
      {integer}
      <span className={styles.fractional}>{fractional}</span>
    </>
  ) : (
    value
  );

  return (
    <span className={styles.amount}>
      {formatted} <KiltCurrency />
    </span>
  );
}
