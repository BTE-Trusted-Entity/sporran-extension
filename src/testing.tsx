export * from '@testing-library/react';
export { waitForElementToBeRemoved } from '@testing-library/dom';
import { act, render as externalRender, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import dialogPolyfill from 'dialog-polyfill';

import { AccountsProviderMock } from './testing/AccountsProviderMock';

export {
  AccountsProviderMock,
  accountsMock,
} from './testing/AccountsProviderMock';

export { mockBackgroundScript } from './testing/mockBackgroundScript';

export function render(
  ui: Parameters<typeof externalRender>[0],
  options?: Parameters<typeof externalRender>[1],
): ReturnType<typeof externalRender> {
  return externalRender(
    <AccountsProviderMock>
      <MemoryRouter>{ui}</MemoryRouter>
    </AccountsProviderMock>,
    options,
  );
}

export async function waitForTooltipUpdate(): Promise<void> {
  expect(await screen.findByRole('status')).toHaveAttribute(
    'data-popper-escaped',
  );
}

const dialogPromise = Promise.resolve();
(dialogPolyfill.registerDialog as jest.Mock).mockReturnValue(dialogPromise);

export async function waitForDialogUpdate(): Promise<void> {
  await act(() => dialogPromise);
}
