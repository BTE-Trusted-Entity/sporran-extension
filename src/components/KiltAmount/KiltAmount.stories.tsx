import { Meta } from '@storybook/react';
import { BalanceUtils } from '@kiltprotocol/sdk-js';

import { KiltAmount } from './KiltAmount';

export default {
  title: 'Components/KiltAmount',
  component: KiltAmount,
} as Meta;

export function Funds(): JSX.Element {
  return (
    <KiltAmount amount={BalanceUtils.toFemtoKilt(1.23459999)} type="funds" />
  );
}

export function Costs(): JSX.Element {
  return (
    <KiltAmount amount={BalanceUtils.toFemtoKilt(1.23459999)} type="costs" />
  );
}
