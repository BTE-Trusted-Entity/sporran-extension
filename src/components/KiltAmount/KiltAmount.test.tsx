import { BalanceUtils } from '@kiltprotocol/sdk-js';

import { render } from '../../testing/testing';

import { asKiltCoins, KiltAmount } from './KiltAmount';

describe('KiltAmount', () => {
  describe('asKiltCoins', () => {
    it('should stringify the BN', async () => {
      expect(asKiltCoins(BalanceUtils.toFemtoKilt(1.234), 'funds')).toEqual(
        '1.2340',
      );
    });

    describe('funds', () => {
      it('should keep value when no rounding is needed', async () => {
        expect(asKiltCoins(BalanceUtils.toFemtoKilt(1.2345), 'funds')).toEqual(
          '1.2345',
        );
      });

      it('should round down the slightly larger value', async () => {
        expect(
          asKiltCoins(BalanceUtils.toFemtoKilt(1.23450001), 'funds'),
        ).toEqual('1.2345');
      });

      it('should round down the almost next value', async () => {
        expect(
          asKiltCoins(BalanceUtils.toFemtoKilt(1.23459999), 'funds'),
        ).toEqual('1.2345');
        expect(
          asKiltCoins(BalanceUtils.toFemtoKilt(1.99999999), 'funds'),
        ).toEqual('1.9999');
      });
    });

    describe('costs', () => {
      it('should keep value when no rounding is needed', async () => {
        expect(asKiltCoins(BalanceUtils.toFemtoKilt(1.2345), 'costs')).toEqual(
          '1.2345',
        );
      });

      it('should round up even the slightly larger values', async () => {
        expect(
          asKiltCoins(BalanceUtils.toFemtoKilt(1.00000001), 'costs'),
        ).toEqual('1.0001');
        expect(
          asKiltCoins(BalanceUtils.toFemtoKilt(1.23450001), 'costs'),
        ).toEqual('1.2346');
      });

      it('should round up the almost next value', async () => {
        expect(
          asKiltCoins(BalanceUtils.toFemtoKilt(1.23450001), 'costs'),
        ).toEqual('1.2346');
      });
    });
  });

  describe('KiltAmount', () => {
    it('should render funds', async () => {
      const { container } = render(
        <KiltAmount
          amount={BalanceUtils.toFemtoKilt(1.23459999)}
          type="funds"
        />,
      );
      expect(container).toMatchSnapshot();
    });
    it('should render costs', async () => {
      const { container } = render(
        <KiltAmount
          amount={BalanceUtils.toFemtoKilt(1.23450001)}
          type="costs"
        />,
      );
      expect(container).toMatchSnapshot();
    });
    it('should render with small fractional part', async () => {
      const { container } = render(
        <KiltAmount
          amount={BalanceUtils.toFemtoKilt(1.23450001)}
          type="costs"
          smallDecimals
        />,
      );
      expect(container).toMatchSnapshot();
    });
  });
});
