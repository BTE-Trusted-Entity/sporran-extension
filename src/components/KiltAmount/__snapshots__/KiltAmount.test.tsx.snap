// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KiltAmount KiltAmount should render costs 1`] = `
<div>
  <span
    class="amount"
  >
    1.2346
     
    <span
      aria-label="KILT Coin"
      class="coin"
    />
  </span>
</div>
`;

exports[`KiltAmount KiltAmount should render funds 1`] = `
<div>
  <span
    class="amount"
  >
    1.2345
     
    <span
      aria-label="KILT Coin"
      class="coin"
    />
  </span>
</div>
`;

exports[`KiltAmount KiltAmount should render with small fractional part 1`] = `
<div>
  <span
    class="amount"
  >
    1
    <span
      class="fractional"
    >
      .2346
    </span>
     
    <span
      aria-label="KILT Coin"
      class="coin"
    />
  </span>
</div>
`;
